src/StreamingDataFrames/streamingdataframes/state/rocksdb/stores.py

```diff
@@ -0,0 +1,432 @@
+
+    Responsibilities:
+     1. Managing access to the RocksDB instance
+     2. Creating transactions to interact with data
+     3. Flushing WriteBatches to the RocksDB
+
+    It opens the RocksDB on `__init__`. If the db is locked by another process,
+    it will retry opening according to `open_max_retries` and `open_retry_backoff`
+
+    :param path: an absolute path to the RocksDB folder
+    :param options: RocksDB options. If `None`, the default options will be used.
+    :param open_max_retries: number of times to retry opening the database
+        if it's locked by another process. To disable retrying, pass 0.
+    :param open_retry_backoff: number of seconds to wait between each retry.
+    :param dumps: the function used to serialize keys & values to bytes in transactions.
+        Default - `json.dumps`
+    :param loads: the function used to deserialize keys & values from bytes to objects
+        in transactions. Default - `json.loads`.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        options: Optional[RocksDBOptionsProto] = None,
+        open_max_retries: int = 10,
+        open_retry_backoff: float = 3.0,
+        dumps: Optional[DumpsFunc] = None,
+        loads: Optional[LoadsFunc] = None,
+    ):
+        self._path = path
+        self._options = options or RocksDBOptions()
+        self._open_max_retries = open_max_retries
+        self._open_retry_backoff = open_retry_backoff
+        self._dumps = dumps
+        self._loads = loads
+        self._db = self._open_db()
+
+    def begin(self) -> "TransactionStore":
+        """
+        Create a new `TransactionStore` object.
+        Using `TransactionStore` is a recommended way for accessing the data.
+
+        :return: an instance of `TransactionStore`
+        """
+        return TransactionStore(storage=self, dumps=self._dumps, loads=self._loads)
+
+    def write(self, batch: rocksdict.WriteBatch):
+        """
+        Write `WriteBatch` to RocksDB
+        :param batch: an instance of `rocksdict.WriteBatch`
+        """
+        self._db.write(batch)
+
+    def get(self, key: bytes, default: Any = None) -> Union[None, bytes, Any]:
+        """
+        Get a key from RocksDB.
+
+        :param key: a key encoded to `bytes`
+        :param default: a default value to return if the key is not found.
+        :return: a value if the key is present in the DB. Otherwise, `None` or `default`
+        """
+        return self._db.get(key, default)
+
+    def exists(self, key: bytes) -> bool:
+        """
+        Check if a key is present in the DB.
+
+        :param key: a key encoded to `bytes`.
+        :return: `True` if the key is present, `False` otherwise.
+        """
+        return key in self._db
+
+    def flush(self, wait: bool = False):
+        """
+        Manually flush the current mem-table.
+
+        :param wait: whether to wait for the flush to finish
+        """
+        self._db.flush(wait)
+
+    def close(self):
```

Comment:
Is flushing upon closing something you need/should do? Or is that handled automatically, frowned upon, or just not something we want to do?

Reply:
Nice find. 
The `flush` function actually contains a mistake here (it doesn't expect any arguments).
And we don't really call it anywhere right now, `Rdict.close()` already flushes the current memtable inside.
I'll remove it.